File: benchmark/languages/python/track_a/fasta.py
#!/usr/bin/env python3
import sys

SOURCE = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"
WEIGHTS = (
    ("A", 270),
    ("C", 120),
    ("G", 120),
    ("T", 270),
    ("N", 220),
)


def count_repeat(length: int) -> dict[str, int]:
    counts = {key: 0 for key, _ in WEIGHTS}
    for i in range(length):
        base = SOURCE[i % len(SOURCE)]
        counts[base] += 1
    return counts


def choose_base(value: int) -> str:
    total = 0
    for base, weight in WEIGHTS:
        total += weight
        if value < total:
            return base
    return "N"


def count_random(length: int, seed: int) -> dict[str, int]:
    counts = {key: 0 for key, _ in WEIGHTS}
    state = seed
    for _ in range(length):
        state = (state * 3877 + 29573) % 139968
        base = choose_base(state % 1000)
        counts[base] += 1
    return counts


def main() -> int:
    repeat_length = int(sys.argv[1]) if len(sys.argv) > 1 else 4000
    random_length = int(sys.argv[2]) if len(sys.argv) > 2 else 4000
    seed = int(sys.argv[3]) if len(sys.argv) > 3 else 42

    repeat_counts = count_repeat(repeat_length)
    random_counts = count_random(random_length, seed)

    print(repeat_length)
    print(random_length)
    print(repeat_counts["A"])
    print(repeat_counts["C"])
    print(repeat_counts["G"])
    print(repeat_counts["T"])
    print(random_counts["A"])
    print(random_counts["C"])
    print(random_counts["G"])
    print(random_counts["T"])
    print(random_counts["N"])
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmark/languages/go/track_a/fasta.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

const source = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"

var orderedBases = []byte{'A', 'C', 'G', 'T', 'N'}
var baseWeights = []int{270, 120, 120, 270, 220}

func countRepeat(length int) map[byte]int {
	counts := map[byte]int{'A': 0, 'C': 0, 'G': 0, 'T': 0, 'N': 0}
	for i := 0; i < length; i++ {
		counts[source[i%len(source)]]++
	}
	return counts
}

func chooseBase(value int) byte {
	total := 0
	for i, weight := range baseWeights {
		total += weight
		if value < total {
			return orderedBases[i]
		}
	}
	return 'N'
}

func countRandom(length int, seed int) map[byte]int {
	counts := map[byte]int{'A': 0, 'C': 0, 'G': 0, 'T': 0, 'N': 0}
	state := seed
	for i := 0; i < length; i++ {
		state = (state*3877 + 29573) % 139968
		counts[chooseBase(state%1000)]++
	}
	return counts
}

func main() {
	repeatLength := 4000
	randomLength := 4000
	seed := 42
	if len(os.Args) > 1 {
		if parsed, err := strconv.Atoi(os.Args[1]); err == nil {
			repeatLength = parsed
		}
	}
	if len(os.Args) > 2 {
		if parsed, err := strconv.Atoi(os.Args[2]); err == nil {
			randomLength = parsed
		}
	}
	if len(os.Args) > 3 {
		if parsed, err := strconv.Atoi(os.Args[3]); err == nil {
			seed = parsed
		}
	}

	repeatCounts := countRepeat(repeatLength)
	randomCounts := countRandom(randomLength, seed)

	fmt.Println(repeatLength)
	fmt.Println(randomLength)
	fmt.Println(repeatCounts['A'])
	fmt.Println(repeatCounts['C'])
	fmt.Println(repeatCounts['G'])
	fmt.Println(repeatCounts['T'])
	fmt.Println(randomCounts['A'])
	fmt.Println(randomCounts['C'])
	fmt.Println(randomCounts['G'])
	fmt.Println(randomCounts['T'])
	fmt.Println(randomCounts['N'])
}
